File: Awful.apk/src/main/java/com/ferg/awfulapp/task/BookmarkColorRequest.java
package com.ferg.awfulapp.task;

import android.content.Context;
import android.net.Uri;

import com.ferg.awfulapp.constants.Constants;
import com.ferg.awfulapp.util.AwfulError;

import org.jsoup.nodes.Document;

/**
 * Created by Matthew on 8/9/13.
 */
public class BookmarkColorRequest extends AwfulRequest<Void> {
    public BookmarkColorRequest(Context context, int threadId) {
        super(context, null);
        addPostParam(Constants.PARAM_ACTION, "cat_toggle");
        addPostParam(Constants.PARAM_THREAD_ID, Integer.toString(threadId));
    }

    @Override
    protected String generateUrl(Uri.Builder urlBuilder) {
        //since we aren't adding query arguments to a POST request,
        //we can just pass null in the constructor URL field and it'll skip this Uri.Builder
        return Constants.FUNCTION_BOOKMARK;
    }

    @Override
    protected Void handleResponse(Document doc) throws AwfulError {
        //nothing to do
        return null;
    }

    @Override
    protected boolean handleError(AwfulError error, Document doc) {
        return error.isCritical();
    }

}

File: Awful.apk/src/main/java/com/ferg/awfulapp/task/BookmarkColorRequest.kt
package com.ferg.awfulapp.task

import android.content.Context
import com.ferg.awfulapp.constants.Constants.*
import com.ferg.awfulapp.util.AwfulError
import org.jsoup.nodes.Document

/**
 * Cycle the bookmark colour set on the site for a given thread.
 */
class BookmarkColorRequest(context: Context, threadId: Int)
    : AwfulRequest<Void?>(context, FUNCTION_BOOKMARK, isPostRequest = true) {
    init {
        with (parameters) {
            add(PARAM_ACTION, "cat_toggle")
            add(PARAM_THREAD_ID, Integer.toString(threadId))
        }
    }

    @Throws(AwfulError::class)
    override fun handleResponse(doc: Document): Void? = null

}
